server/grpc_service.go

```diff
@@ -233,8 +234,20 @@ func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb.StoreHeartbea
 	}, nil
 }
 
+type syncHeartbeatServer struct {
+	sync.Mutex
+	server pdpb.PD_RegionHeartbeatServer
+}
+
+func (s *syncHeartbeatServer) Send(msg *pdpb.RegionHeartbeatResponse) error {
+	s.Lock()
+	defer s.Unlock()
+	return s.server.Send(msg)
+}
+
 // RegionHeartbeat implements gRPC PDServer.
 func (s *Server) RegionHeartbeat(server pdpb.PD_RegionHeartbeatServer) error {
+	syncServer := &syncHeartbeatServer{server: server}
```

Comment:
I think send the error with hbstreams same as a normal response better than use lock.